tests/cases/compiler/narrowingDestructuring.ts

```diff
@@ -0,0 +1,39 @@
+type X = { kind: "a", a: string } | { kind: "b", b: string }
+
+function func<T extends X>(value: T) {
+    if (value.kind === "a") {
+        value.a;
+        const { a } = value;
+    } else {
+        value.b;
+        const { b } = value;
+    }
+}
+
+type Z = { kind: "f", f: { a: number, b: string, c: number } }
+    | { kind: "g", g: { a: string, b: number, c: string }};
+
+function func2<T extends Z>(value: T) {
+    if (value.kind === "f") {
+        const { f: f1 } = value;
+        const { f: { a, ...spread } } = value;
+        value.f;
+    } else {
+        const { g: { c, ...spread } } = value;
+        value.g;
+    }
+}
+
+function func3<T extends { kind: "a", a: string } | { kind: "b", b: number }>(t: T) {
+    if (t.kind === "a") {
+        const { kind, ...r1 } = t;
+        const r2 = (({ kind, ...rest }) => rest)(t);
```

Comment:
That's an inconsistency that was already present. We have:
`r1 : Omit<T, "kind">`, but `r2 : { a: string; }`.